In [1]:
# ---------------- #
# Common Libraries #
# ---------------- #
      
# Standard Imports
import os
import re
import nltk
import requests
import unicodedata
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from nltk.corpus import stopwords

## setting basic style parameters for matplotlib
plt.rc('figure', figsize=(13, 7))
plt.style.use('seaborn-darkgrid')

# ------------- #
# LOCAL IMPORTS #
# ------------- #

# importing sys
import sys
# adding 00_helper_files to the system path
sys.path.insert(0, '/Users/qmcbt/codeup-data-science/00_helper_files')
# env containing sensitive access credentials
import env
from env import user, password, host
from env import get_db_url

# Import Local Helper Modules
import QMCBT_00_quicktips as qt
import QMCBT_01_acquire as acq
import QMCBT_02_prepare as prep
import QMCBT_03_explore as exp
import QMCBT_04_visualize as viz
import QMCBT_05_model as mod
import QMCBT_wrangle as w

# Import py modules
import prepare as p

In [2]:
# Read-in data from compressed file
df = pd.read_pickle('michelin_df.pickle')

In [3]:
# Display DataFrame
df.T.head(20)

0     \
name                                                            King's Joy   
address                        2 Wudaoying Hutong, Beijing, China Mainland   
location                                           Beijing, China Mainland   
price                                                                  ¥¥¥   
cuisine                                                         Vegetarian   
longitude                                                       116.410004   
latitude                                                         39.946681   
phone_number                                                861084049191.0   
url                      https://guide.michelin.com/en/beijing-municipa...   
website_url                                                            NaN   
award                                                     3 MICHELIN Stars   
facilities_and_services  Air conditioning,American Express credit card,...   
data                     Just a stone’s throw from Yonghe Temple, this ...   

                                                                      1     \
name                                      Xin Rong Ji (Xinyuan South Road)   
address                  1F, East Tower, Genesis Beijing, 8 Xinyuan Sou...   
location                                           Beijing, China Mainland   
price                                                                 ¥¥¥¥   
cuisine                                                            Taizhou   
longitude                                                       116.450148   
latitude                                                          39.94638   
phone_number                                                861065015501.0   
url                      https://guide.michelin.com/en/beijing-municipa...   
website_url                                                            NaN   
award                                                     3 MICHELIN Stars   
facilities_and_services  Air conditioning,Car park,China UnionPay,Count...   
data                     This branch of the chain restaurant opened in ...   

                                                                      2     \
name                                                           Taian Table   
address                  101-102, Building No. 1, Garden Office, No.161...   
location                                          Shanghai, China Mainland   
price                                                                 ¥¥¥¥   
cuisine                                                         Innovative   
longitude                                                       121.474049   
latitude                                                         31.221807   
phone_number                                               8617301605350.0   
url                      https://guide.michelin.com/en/shanghai-municip...   
website_url                                        https://taian-table.cn/   
award                                                     3 MICHELIN Stars   
facilities_and_services  Air conditioning,American Express credit card,...   
data                     A fixture on the city’s dining scene since 201...   

                                                                      3     \
name                                            Ultraviolet by Paul Pairet   
address                  'somewhere in Shanghai' - meet at Mr & Mrs Bun...   
location                                          Shanghai, China Mainland   
price                                                                 ¥¥¥¥   
cuisine                                                         Innovative   
longitude                                                        121.48509   
latitude                                                         31.240358   
phone_number                                                           NaN   
url                      https://guide.michelin.com/en/shanghai-municip...   
website_url

In [4]:
train, validate, test = p.prepare_michelin(df)

In [5]:
train.T.head(20)

3016  \
name                                                     rüssel's landhaus   
address                     büdlicherbrück 1, naurath/wald, 54426, germany   
location                                             naurath/wald, germany   
price                                                                 €€€€   
cuisine                                           creative, market cuisine   
longitude                                                         6.892007   
latitude                                                         49.768126   
url                      https://guide.michelin.com/en/rheinland-pfalz/...   
award                                                      1 michelin star   
facilities_and_services  american express credit card,car park,credit c...   
data                     the charm of this country house makes itself f...   
price_level                                                              4   
city                                                          naurath/wald   
country                                                            germany   
clean                    charm country house makes felt approach via sm...   
lemmatized               charm country house make felt approach via sma...   

                                                                      1802  \
name                                                                 iroha   
address                  5f, 1-3-19 sonezakishinchi, kita-ku, osaka, 53...   
location                                                      osaka, japan   
price                                                                  ¥¥¥   
cuisine                                                           japanese   
longitude                                                       135.495812   
latitude                                                         34.697842   
url                      https://guide.michelin.com/en/osaka-region/osa...   
award                                                      1 michelin star   
facilities_and_services  air conditioning,american express credit card,...   
data                     the charm here is in the wide selection of dis...   
price_level                                                              3   
city                                                                 osaka   
country                                                              japan   
clean                    charm wide selection dishes consists classic d...   
lemmatized               charm wide selection dish consists classic dis...   

                                                                      91    \
name                                                                   kei   
address                           5 rue du coq-héron, paris, 75001, france   
location                                                     paris, france   
price                                                                 €€€€   
cuisine                                                     modern cuisine   
longitude                                                         2.342285   
latitude                                                         48.864395   
url                      https://guide.michelin.com/en/ile-de-france/pa...   
award                                                     3 michelin stars   
facilities_and_services  air conditioning,american express credit card,...   
data                     "kei" is kei kobayashi, a chef born in the jap...   
price_level                                                              4   
city                                                                 paris   
country                                                             france   
clean                    kei kei kobayashi born japanese city nagano tr...   
lemmatized               kei kei kobayashi born japanese city nagano tr...   

                                                                      5283  \
name       

In [6]:
# Display DataFrame information
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3796 entries, 3016 to 1195
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   name                     3796 non-null   string 
 1   address                  3796 non-null   string 
 2   location                 3796 non-null   string 
 3   price                    3796 non-null   object 
 4   cuisine                  3796 non-null   string 
 5   longitude                3796 non-null   float64
 6   latitude                 3796 non-null   float64
 7   url                      3796 non-null   object 
 8   award                    3796 non-null   string 
 9   facilities_and_services  3796 non-null   string 
 10  data                     3796 non-null   string 
 11  price_level              3796 non-null   int64  
 12  city                     3796 non-null   object 
 13  country                  3796 non-null   object 
 14  clean                

<div class="alert alert-info">

# Observations
* Dropping all nulls would remove 1336 documents (~135 documents have >1 Null)  

                          price:   54 Nulls
                   phone_number:  131 Nulls
                    website_url: 1238 Nulls
        facilities_and_services:   48 Nulls

* facilities_and_services nunique is 1571 

        * there are only 120 documents that share values with 10 or more other documents
        * there are 540 documents that share values with 2-9 other documents
        * there are 910 documents that are entirely unique

* There are only 16 locations that have more than 50 documents (awardees) and only half of those contain more than 100 documents  

        * Tokyo, Japan                    422
        * Kyoto, Japan                    196
        * Osaka, Japan                    193
        * New York, USA                   177
        * Paris, France                   164
        * Hong Kong                       135
        * Singapore                       119
        * London, United Kingdom          104
        * Bangkok, Thailand               103

* There are 863 unique cuisine values; 26 cuisine types that have more than 50 documents and only 9 that have more than 100 documents; 445 unique cuisine type combinations have only one document.  
        
        * Modern Cuisine              994
        * Creative                    416
        * Japanese                    287
        * Traditional Cuisine         218
        * French                      184
        * Street Food                 157
        * Italian                     130
        * Contemporary                114
        * Cantonese                   112

* There are four different award types  

        * Bib Gourmand        3411
        * 1 MICHELIN Star     2744
        * 2 MICHELIN Stars     483
        * 3 MICHELIN Stars     142

# Clean
### DTYPE & IMPUTE
* Cast facilities_and_services astype('string') and fillna('NONE')
* Cast data astype('string')

# Feature Engineer
* NON-MVP Option - Create buckets and encode locations and cuisine
* Encode award 
* Use facilities_and_services as a feature
* IMPUTE prce with len and encode


In [7]:
# Display count of unique entries
train.facilities_and_services.nunique()

1127

In [8]:
# Show aggregated value counts
train.facilities_and_services.value_counts().head(435)

air conditioning,american express credit card,...    149
cash only                                             88
air conditioning,credit card / debit card acce...     81
air conditioning,american express credit card,...     73
credit card / debit card accepted,mastercard c...     72
                                                    ... 
credit card / debit card accepted,mastercard c...      2
american express credit card,credit card / deb...      2
particularly interesting wine list,valet parking       2
air conditioning,american express credit card,...      1
air conditioning,american express credit card,...      1
Name: facilities_and_services, Length: 435, dtype: Int64

## Award & Price Comparisons 

In [9]:
# Award Counts
train.award.value_counts()

bib gourmand        1910
1 michelin star     1536
2 michelin stars     270
3 michelin stars      80
Name: award, dtype: Int64

In [10]:
# Award Percentages
print(f'Bib Gourmand: {1910/3796}')
print(f'      1-Star: {1536/3796}')
print(f'      2-Star: {270/3796}')
print(f'      3-Star: {80/3796}')

Bib Gourmand: 0.5031612223393045
      1-Star: 0.40463645943098
      2-Star: 0.07112750263435196
      3-Star: 0.02107481559536354


In [11]:
# Price Count
train.price_level.value_counts()

2    1373
4    1025
3     815
1     583
Name: price_level, dtype: int64

In [12]:
# Price Percentages
print(f'Price Level 1: {583/3796}')
print(f'Price Level 2: {1373/3796}')
print(f'Price Level 3: {815/3796}')
print(f'Price Level 4: {1025/3796}')

Price Level 1: 0.1535827186512118
Price Level 2: 0.3616965226554268
Price Level 3: 0.21469968387776606
Price Level 4: 0.27002107481559534


#### Bib Gourmand

In [13]:
# Bib Gourmand Price Count
train.price_level[train.award == 'bib gourmand'].value_counts()

2    1300
1     580
3      29
4       1
Name: price_level, dtype: int64

In [14]:
# Price Percentages
print(f'Bib Gourmand Price Level 1: {1300/1373}')
print(f'Bib Gourmand Price Level 2: {580/583}')
print(f'Bib Gourmand Price Level 3: {29/815}')
print(f'Bib Gourmand Price Level 4: {1/1025}')
print('------------------------------------------------------------------')
print(f'Bib Gourmand Percentage of Total: {(1300+580+29+1)/(1373+583+815+1025)}')

Bib Gourmand Price Level 1: 0.9468317552804079
Bib Gourmand Price Level 2: 0.9948542024013722
Bib Gourmand Price Level 3: 0.03558282208588957
Bib Gourmand Price Level 4: 0.000975609756097561
------------------------------------------------------------------
Bib Gourmand Percentage of Total: 0.5031612223393045


#### 1-Star

In [15]:
# Price Count
train.price_level[train.award == '1 michelin star'].value_counts()

3    747
4    719
2     67
1      3
Name: price_level, dtype: int64

In [16]:
# Price Percentages
print(f'1 Star Price Level 1: {3/1373}')
print(f'1 Star Price Level 2: {67/583}')
print(f'1 Star Price Level 3: {747/815}')
print(f'1 Star Price Level 4: {719/1025}')
print('------------------------------------------------------------------')
print(f'1 Star Percentage of Total: {(3+67+747+719)/(1373+583+815+1025)}')

1 Star Price Level 1: 0.0021849963583394027
1 Star Price Level 2: 0.11492281303602059
1 Star Price Level 3: 0.9165644171779141
1 Star Price Level 4: 0.7014634146341463
------------------------------------------------------------------
1 Star Percentage of Total: 0.40463645943098


#### 2-Star

In [17]:
# Price Count
train.price_level[train.award == '2 michelin stars'].value_counts()

4    229
3     35
2      6
Name: price_level, dtype: int64

In [18]:
# Price Percentages
print(f'2 Star Price Level 1: {0/1373}')
print(f'2 Star Price Level 2: {6/583}')
print(f'2 Star Price Level 3: {35/815}')
print(f'2 Star Price Level 4: {229/1025}')
print('------------------------------------------------------------------')
print(f'2 Star Percentage of Total: {(0+6+35+229)/(1373+583+815+1025)}')

2 Star Price Level 1: 0.0
2 Star Price Level 2: 0.010291595197255575
2 Star Price Level 3: 0.04294478527607362
2 Star Price Level 4: 0.22341463414634147
------------------------------------------------------------------
2 Star Percentage of Total: 0.07112750263435196


#### 3-Star

In [19]:
# Price Count
train.price_level[train.award == '3 michelin stars'].value_counts()

4    76
3     4
Name: price_level, dtype: int64

In [20]:
# Price Percentages
print(f'3 Star Price Level 1: {0/1373}')
print(f'3 Star Price Level 2: {0/583}')
print(f'3 Star Price Level 3: {4/815}')
print(f'3 Star Price Level 4: {76/1025}')
print('------------------------------------------------------------------')
print(f'3 Star Percentage of Total: {(0+0+4+76)/(1373+583+815+1025)}')

3 Star Price Level 1: 0.0
3 Star Price Level 2: 0.0
3 Star Price Level 3: 0.0049079754601227
3 Star Price Level 4: 0.07414634146341463
------------------------------------------------------------------
3 Star Percentage of Total: 0.02107481559536354


## Cuisine

In [21]:
train.cuisine.nunique()

655

In [22]:
# cuisine counts > 50
train.cuisine.value_counts().head(11)

modern cuisine         552
creative               216
japanese               142
traditional cuisine    114
french                 102
street food             98
italian                 73
contemporary            65
cantonese               57
classic cuisine         51
modern french           46
Name: cuisine, dtype: Int64

## Locations

In [23]:
# Unique of locations
train.location.nunique()

1673

In [24]:
# location counts > 50
train.location.value_counts().head(11)

tokyo, japan              202
osaka, japan              116
new york, usa             107
kyoto, japan               96
paris, france              85
hong kong                  75
singapore                  70
bangkok, thailand          56
seoul, south korea         55
london, united kingdom     53
chicago, usa               49
Name: location, dtype: Int64

In [25]:
# Country Count
train.country.nunique()

40

In [26]:
# Country List
train.country.unique()

array(['germany', 'japan', 'france', 'singapore', 'italy',
       'taipei & taichung', 'switzerland', 'thailand', 'china mainland',
       'hong kong', 'spain', 'usa', 'greece', 'south korea',
       'united kingdom', 'norway', 'portugal', 'dubai', 'netherlands',
       'malaysia', 'hungary', 'ireland', 'brazil', 'belgium', 'austria',
       'denmark', 'croatia', 'canada', 'finland', 'macau',
       'czech republic', 'malta', 'türkiye', 'slovenia', 'luxembourg',
       'sweden', 'abu dhabi', 'poland', 'estonia', 'andorra'],
      dtype=object)

In [27]:
# Award Count by Country >50
train.country.value_counts().head(16)

france               584
japan                434
usa                  364
italy                356
germany              330
spain                282
switzerland          156
united kingdom       148
belgium              145
netherlands          136
thailand             133
china mainland       119
taipei & taichung     93
hong kong             75
singapore             70
south korea           55
Name: country, dtype: int64

#### 1. France (584 awards)

In [28]:
# Award Count by Country
train.award[train.country == 'france'].value_counts()

1 michelin star     301
bib gourmand        225
2 michelin stars     43
3 michelin stars     15
Name: award, dtype: Int64

In [29]:
# Price Count by Country
train.price_level[train.country == 'france'].value_counts()

4    207
2    182
3    152
1     43
Name: price_level, dtype: int64

In [30]:
# Cuisine Count by Country
train.cuisine[train.country == 'france'].value_counts().head()

modern cuisine              307
creative                     86
traditional cuisine          38
classic cuisine              20
creative, modern cuisine     15
Name: cuisine, dtype: Int64

#### 2. Japan (434 awards)

In [31]:
# Award Count by Country
train.award[train.country == 'japan'].value_counts()

bib gourmand        228
1 michelin star     158
2 michelin stars     40
3 michelin stars      8
Name: award, dtype: Int64

In [32]:
# Price Count by Country
train.price_level[train.country == 'japan'].value_counts()

3    155
2    138
1     98
4     43
Name: price_level, dtype: int64

In [33]:
# Cuisine Count by Country
train.cuisine[train.country == 'japan'].value_counts().head()

japanese    104
french       72
sushi        31
italian      23
izakaya      21
Name: cuisine, dtype: Int64

#### 3. USA (364 awards)

In [34]:
# Award Count by Country
train.award[train.country == 'usa'].value_counts()

bib gourmand        233
1 michelin star     103
2 michelin stars     19
3 michelin stars      9
Name: award, dtype: Int64

In [35]:
# Price Count by Country
train.price_level[train.country == 'usa'].value_counts()

2    192
4    116
1     33
3     23
Name: price_level, dtype: int64

In [36]:
# Cuisine Count by Country
train.cuisine[train.country == 'usa'].value_counts().head()

japanese, sushi              25
contemporary                 19
mexican                      18
mexican, regional cuisine    12
californian                  10
Name: cuisine, dtype: Int64

#### 4. Italy (356 awards)

In [37]:
# Award Count by Country
train.award[train.country == 'italy'].value_counts()

1 michelin star     194
bib gourmand        138
2 michelin stars     15
3 michelin stars      9
Name: award, dtype: Int64

In [38]:
# Price Count by Country
train.price_level[train.country == 'italy'].value_counts()

4    124
3     91
1     80
2     61
Name: price_level, dtype: int64

In [39]:
# Cuisine Count by Country
train.cuisine[train.country == 'italy'].value_counts().head()

modern cuisine            46
creative                  27
creative, contemporary    21
piedmontese               14
tuscan                    11
Name: cuisine, dtype: Int64

#### 5. Germany (330 awards)

In [40]:
# Award Count by Country
train.award[train.country == 'germany'].value_counts()

1 michelin star     147
bib gourmand        146
2 michelin stars     30
3 michelin stars      7
Name: award, dtype: Int64

In [41]:
# Price Count by Country
train.price_level[train.country == 'germany'].value_counts()

4    139
2    120
3     44
1     27
Name: price_level, dtype: int64

In [42]:
# Cuisine Count by Country
train.cuisine[train.country == 'germany'].value_counts().head()

country cooking, market cuisine    15
creative, modern cuisine           13
modern french                      12
modern cuisine, creative           12
classic french                     11
Name: cuisine, dtype: Int64

### Reviews

In [47]:
# Assign all, 1_star, 2_star, 3_star and bib_gourmand lists by passing the  function with a join
all_reviews = (' '.join(train['lemmatized']))
one_star_reviews = (' '.join(train[train.award == '1 michelin star']['lemmatized']))
two_star_reviews = (' '.join(train[train.award == '2 michelin stars']['lemmatized']))
three_star_reviews = (' '.join(train[train.award == '3 michelin stars']['lemmatized']))
bib_gourmand_reviews = (' '.join(train[train.award == 'bib gourmand']['lemmatized']))

In [50]:
#TODO Create Mean Word Count

In [49]:
# show word count
print('Reviews Word Counts')
print(f'all_reviews: {len(all_reviews)}')
print(f'one_star_reviews: {len(one_star_reviews)}')
print(f'two_star_reviews: {len(two_star_reviews)}')
print(f'three_star_reviews: {len(three_star_reviews)}')
print(f'bib_gourmand_reviews: {len(bib_gourmand_reviews)}')

Reviews Word Counts
all_reviews: 1415666
one_star_reviews: 670142
two_star_reviews: 119009
three_star_reviews: 40599
bib_gourmand_reviews: 585913


In [51]:
# Assign word counts to Frequency Variables
freq_one_star_reviews = pd.Series(one_star_reviews).value_counts()
freq_two_star_reviews = pd.Series(two_star_reviews).value_counts()
freq_three_star_reviews = pd.Series(three_star_reviews).value_counts()
freq_bib_gourmand_reviews = pd.Series(bib_gourmand_reviews).value_counts()
freq_all_reviews = pd.Series(all_reviews).value_counts()

In [52]:
# Test it
freq_all_reviews

charm country house make felt approach via small bridge safe say character idyllic old mill successfully preserved ruth harald russel clearly put heart soul running hotel chefpatron talented kitchen team propose five sevencourse set diner quality ingredient prefer source local producer even hunt game classical foundation always shine elegantly complemented creative flourish never come across fussy unnecessary try surprising wine pairing tend recommend fine moselle wine head terrace sit small lake also appealing guestroom charm wide selection dish consists classic dish new creation including tsukuri seared charcoal beef cutlet served sauce made stewed beef tendon masatoshi kamoto us choice ingredient without regard production area prepares dish suit occasion customer preference enjoy food flavour sense camaraderie prepared kei kei kobayashi born japanese city nagano trained prestigious kitchen threestarred chef gilles goujon l ' auberge du vieux puits fontjoncouse alain ducasse plaza at

### Facilities

In [ ]:
# Assign all, 1_star, 2_star, 3_star and bib_gourmand lists by passing the clean function with a join
all_facilities = clean(' '.join(df['facilities_and_services']),'lem')
one_star_facilities = clean(' '.join(df[df.award == '1 MICHELIN Star']['facilities_and_services']),'lem')
two_star_facilities = clean(' '.join(df[df.award == '2 MICHELIN Stars']['facilities_and_services']),'lem')
three_star_facilities = clean(' '.join(df[df.award == '3 MICHELIN Stars']['facilities_and_services']),'lem')
bib_gourmand_facilities = clean(' '.join(df[df.award == 'Bib Gourmand']['facilities_and_services']),'lem')

In [ ]:
# show word count
print('Facility Word Counts')
print(f'all_facilities: {len(all_facilities)}')
print(f'one_star_facilities: {len(one_star_facilities)}')
print(f'two_star_facilities: {len(two_star_facilities)}')
print(f'three_star_facilities: {len(three_star_facilities)}')
print(f'bib_gourmand_facilities: {len(bib_gourmand_facilities)}')

In [ ]:
# Assign word counts to Frequency Variables
freq_one_star_facilities = pd.Series(one_star_facilities).value_counts()
freq_two_star_facilities = pd.Series(two_star_facilities).value_counts()
freq_three_star_facilities = pd.Series(three_star_facilities).value_counts()
freq_bib_gourmand_facilities = pd.Series(bib_gourmand_facilities).value_counts()
freq_all_facilities = pd.Series(all_facilities).value_counts()

In [ ]:
# Test it
freq_all_facilities

In [ ]:
# Create Frequency DataFrame
word_counts = pd.concat([freq_all_facilities,
                         freq_one_star_facilities, 
                         freq_two_star_facilities,
                         freq_three_star_facilities,
                         freq_bib_gourmand_facilities,
                         freq_all_reviews,
                         freq_one_star_reviews,
                         freq_two_star_reviews,
                         freq_three_star_reviews,
                         freq_bib_gourmand_reviews], axis=1
         ).fillna(0).astype(int)
word_counts.columns = ['all_facilities',
                         '1_star_facilities', 
                         '2_star_facilities',
                         '3_star_facilities',
                         'bib_gourmand_facilities',
                         'all_reviews',
                         '1_star_reviews',
                         '2_star_reviews',
                         '3_star_reviews',
                         'bib_gourmand_reviews']
word_counts